Update tool recommendations (#1474)
pypa/packaging.python.org: source/guides/tool-recommendations.rst

Tool recommendations
====================

The Python packaging landscape consists of many different tools. For many tasks,

> the Python Packaging Authority (PyPA, the umbrella organization which

Reviewer: ```suggestion
the :term:`Python Packaging Authority <Python Packaging Authority (PyPA)>` (PyPA, the umbrella organization which
```

Installing packages
===================

:ref:`Pip` is a standard tool to install packages from :term:`PyPI <Python
Package Index (PyPI)>`. It can install into the global environment or into
virtual environments. You may want to read pip's recommendations for
:doc:`secure installs <pip:topics/secure-installs>`. Pip is available by default
in most Python installations.

Alternatively, for installing Python command line applications specifically,
consider :ref:`pipx`, which is a wrapper around pip that installs each
application into a dedicated virtual environment in order to avoid conflicts
with other applications and, on Linux, conflicts with the system.

For scientific software specifically, consider :ref:`Conda` or :ref:`Spack`.

.. todo:: Write a "pip vs. Conda" comparison, here or in a new discussion.

Do **not** use ``easy_install`` (part of :ref:`setuptools`), which is deprecated
in favor of pip (see :ref:`pip vs easy_install` for details). Likewise, do
**not** use ``python setup.py install`` or ``python setup.py develop``, which
are also deprecated (see :ref:`setup-py-deprecated` for background and
:ref:`modernize-setup-py-project` for migration advice).

> Lock files

Author: Yes, but follow-up PR?

> Lock files

Reviewer: Yup.

==========

:ref:`pip-tools` and :ref:`Pipenv` are two recognized tools to create lock
files, which contain the exact versions of all packages installed into an
environment, for reproducibility purposes.

Build backends
==============

Popular :term:`build backends <build backend>` for pure-Python packages include:

- Hatchling, which is part of :ref:`Hatch` (but can be used without
  Hatch as well). Hatchling is extensible through a plugin system.

- :ref:`setuptools`, the historical build backend. It can be configured
  programmatically through the :file:`setup.py` file.

  If you use setuptools, please be aware that it contains many deprecated
  features which are currently kept for compatibility, but should not be used.

>   For example, do not use ``python setup.py`` invocations

Reviewer: ```suggestion
  For example, do **not** use ``python setup.py`` invocations
```

Tool recommendations
====================

The Python packaging landscape consists of many different tools. For many tasks,
the :term:`Python Packaging Authority <Python Packaging Authority (PyPA)>`

> (PyPA, the umbrella organization which encompasses many packaging tools and

Reviewer 2: I wouldn't say "organization", it's more like a bunch of people and projects that are sometimes connected loosely, and sometimes not.
https://www.pypa.io/ presents the PyPA as a working group, which is probably a better term to use here.

Reviewer 2: Marking this as resolved.

maintains this guide) purposefully does not make a blanket recommendation; for

> example, the reason there exist many build backends is that the landscape was

Reviewer 2: ```suggestion
example, the reason there are many build backends is that the landscape was
```

Reviewer 2: Marking this as resolved.

> opened in order to enable the development of new backends serving certain users'

Reviewer 2: ```suggestion
opened up in order to enable the development of new backends serving certain users'
```

Reviewer 2: Marking this as resolved.

needs better than the previously unique backend, setuptools. This guide does
point to some tools that are widely recognized, and also makes some
recommendations of tools that you should *not* use because they are deprecated
or insecure.

Virtual environments
====================

The standard tools to create and use virtual environments manually are
:ref:`virtualenv` (PyPA project) and :doc:`venv <python:library/venv>` (part of
the Python standard library, though missing some features of virtualenv).

Installing packages
===================

:ref:`Pip` is the standard tool to install packages from :term:`PyPI <Python
Package Index (PyPI)>`. It can install into the global environment or into
virtual environments. You may want to read pip's recommendations for
:doc:`secure installs <pip:topics/secure-installs>`. Pip is available by default

> in most Python installations.

Reviewer 2: ".. through the stdlib package `ensurepip`".

Reviewer 2: Marking this as resolved.

Alternatively, for installing Python command line applications specifically,

> consider :ref:`pipx`, which is a wrapper around pip that installs each

Reviewer 2: "pip and virtualenv"?

Reviewer: > "pip and virtualenv"?

Pip and Venv, actually. Virtualenv might be supported in future versions.

Reviewer 2: Marking this as resolved.

application into a dedicated virtual environment. This avoids conflicts between

> the dependencies of different applications. On Linux, pipx is especially

Reviewer 2: "applications that are distributed as Python package distributions through the PyPI"?

Reviewer 2: Marking this as resolved.

> important because it also avoids conflicts with the system (which are the reason

Reviewer 2: Not sure I'd separate Linux from others here — it's bad to sure globally shared envs in so many contexts. In GNU/Linux you can break some system software, including package managers, but a similar influence exists in other OS-independent situations.

Reviewer 2: Marking this as resolved.

for :ref:`externally managed environments <externally-managed-environments>`).

For scientific software specifically, consider :ref:`Conda` or :ref:`Spack`.

.. todo:: Write a "pip vs. Conda" comparison, here or in a new discussion.

Do **not** use ``easy_install`` (part of :ref:`setuptools`), which is deprecated
in favor of pip (see :ref:`pip vs easy_install` for details). Likewise, do
**not** use ``python setup.py install`` or ``python setup.py develop``, which
are also deprecated (see :ref:`setup-py-deprecated` for background and
:ref:`modernize-setup-py-project` for migration advice).

Lock files
==========

:ref:`pip-tools` and :ref:`Pipenv` are two recognized tools to create lock
files, which contain the exact versions of all packages installed into an
environment, for reproducibility purposes.

Build backends
==============

> Popular :term:`build backends <build backend>` for pure-Python packages include:

Reviewer 3: ```suggestion
Popular :term:`build backends <build backend>` for pure-Python packages include (in alphabetical order):
```

... and sort them.

> - Hatchling, which is part of :ref:`Hatch` (but can be used without

Reviewer 2: It's a part of the same Git repository but is published a fully separate distribution on PyPI. This should either clarify that it's a part of the Hatch ecosystem or just drop the implication.

Hatchling is a build backend, but Hatch is a frontend, installer, resolver, virtualenv manager, command runner, workflow tool. Let's not smash them together.

Reviewer 2: Marking this as resolved.

  Hatch as well). Hatchling is extensible through a plugin system.

> - :ref:`setuptools`, the historical build backend. It can be configured

Reviewer 2: It's not historical. It's old, but it's also modern. It has a lot of legacy in its own codebase but that's an implementation detail.

cc @jaraco @abravalheri

Author: I didn't mean that it shouldn't be used anymore, only that it was the only choice for a long time. I think this is useful to put on this page because many people will be wondering why the heck they didn't have to make any choice of backend when they created a package years ago. Maybe `` :ref:`setuptools`, which used to be the only build backend `` would be clearer?

Reviewer 2: Yep, that would be better. Thanks!

Reviewer 2: Marking this as resolved.

>   programmatically through the :file:`setup.py` file (but for basic metadata,

Reviewer 2: Or `setup.cfg`, if we're listing the configuration details here. But then, why don't we do this for the other list items? If we're listing the config files, why not mention `hatch.toml`, `.pip-tools.toml`, `Pipfile` above too?

I'd drop these from here too.

Author: I'm just trying to list which backends support plugins / programmatic config. (I think this is important, because those that do have an ecosystem of plugins around them.) `pyproject.toml` was originally not mentioned here, but added per https://github.com/pypa/packaging.python.org/pull/1474#discussion_r1438610971

Reviewer 2: Okay, but I still want https://github.com/pypa/packaging.python.org/pull/1474/files#r1466878549.

I'll mark this as resolved in favor of that thread.

  :file:`pyproject.toml` is preferred).

  If you use setuptools, please be aware that it contains many deprecated
  features which are currently kept for compatibility, but should not be used.
  For example, do **not** use ``python setup.py`` invocations

>   (cf. :ref:`setup-py-deprecated`), the ``python_requires`` argument to

Author: ```suggestion
  (cf. :ref:`setup-py-deprecated`), the ``setup_requires`` argument to
```

  ``setup()`` (use the :ref:`[build-system] table
  <pyproject-guide-build-system-table>` of :file:`pyproject.toml` instead), or
  the ``easy_install`` command (cf. :ref:`pip vs easy_install`).

> - Flit-core, part of :ref:`Flit` (but usable standalone). It is meant to be a

Reviewer 2: Same as with hatch — let's not mix self-sufficient front ends into a section about backends.

Reviewer 2: Marking this as resolved.

  minimal and opinionated build backend. It is not extensible.

- PDM-backend_, part of :ref:`PDM` (but usable standalone). It provides build
  hooks for extensibility.

- Poetry-core, part of :ref:`Poetry` (but usable standalone). It is extensible
  through plugins.

Do **not** use :ref:`distutils`, which is deprecated, and has been removed from
the standard library in Python 3.12, although it still remains available from
setuptools.

For packages with :term:`extension modules <extension module>`, you may use
setuptools, but consider using a build system dedicated to the language the

> extension is written in, such as Meson or CMake for C/C++, or Cargo for Rust,

Reviewer 3: Just want to point out that meson and cmake aren't specific to C/C++

They also support C#, Cuda, Swift, Asm, ObjC, ObjC++, Fortran.
Meson supports D, rust, vala, Java, and cython. CMake supports Hip and ISPC.

They are particularly valuable in cases where you have multiple languages, e.g. a mixture of rust and C++ and Fortran in a single project.

Meson also explicitly allows you to select `*.py files for installation`, do code generation of .py files (and other files), compile data such as gettext .po -> .mo catalogs, produce bundled `*.exe`s, and in general any form of custom build step. In theory CMake allows you to do precisely the same thing, although the scikit-build(-core) docs don't clearly indicate whether they support packaging that into a wheel.

tl;dr saying to use it "to support a dedicated extension language" is vastly underselling them. Even the "C/C++" part is simply *factually* wrong, if you need C/C++ support then setuptools works okay. The killer feature in the data science community is support for *Fortran*.

Reviewer 2: @eli-schwartz are you okay with how it's been updated after your comment?

and bridging this build system to Python using a dedicated build backend:

- :ref:`meson-python` for Meson,

- :ref:`scikit-build-core` for CMake,

- :ref:`maturin` for Cargo.

Building distributions
======================

The standard tool to build :term:`source distributions <source distribution (or
"sdist")>` and :term:`wheels <wheel>` for uploading to PyPI is :ref:`build`.  It
will invoke whichever build backend you :ref:`declared
<pyproject-guide-build-system-table>` in :file:`pyproject.toml`.

Do **not** use ``python setup.py sdist`` and ``python setup.py bdist_wheel`` for
this task. All direct invocations of :file:`setup.py` are :ref:`deprecated
<setup-py-deprecated>`.

If you have :term:`extension modules <extension module>` and want to distribute
wheels for multiple platforms, use :ref:`cibuildwheel` as part of your CI setup

> to build distributable wheels.

Reviewer 2: Let's mention `pip wheel` that can be used as a low-level piece of automation for building platform-specific wheels in places where `cibuildwheel` may be unavailable. With a note that pypa/build can be used for the same.

Author: I thought the pip developers were unhappy with `pip wheel` and considering to deprecate it (but I may be misremembering, and I can't find the pip issue right now). @pfmoore @pradyunsg What's your opinion?

Reviewer 4: I'm not aware of plans to deprecate `pip wheel`, but it's worth noting that:

1. *Building* wheels isn't its sole/primary purpose - it downloads wheels, or downloads sdists and builds them into wheels. `pip wheel .` builds a wheel of the current project, but somewhat as a side issue (we could potentially use a cached wheel if we had a way to be sure it matched the source code).
2. There's no pip command to build a sdist, and there may well never be one.

If you want to build a local project, `build` is a much more appropriate tool.

Author: Ok, then I'd rather not mention `pip wheel` here.

Reviewer 2: Marking this as resolved.

Uploading to PyPI
=================

> The standard tool for this task is :ref:`twine`.

Reviewer 2: Let's mention that some of the workflow tools (hatch, poetry, flit) invoke it under the hood.

Also, let's link the secretless publishing guide which wraps it but should probably be the first choice for any GitHub-hosted projects using GHA: https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/

Author: > Let's mention that some of the workflow tools (hatch, poetry, flit) invoke it under the hood.

I don't think they do? I didn't find any trace of this in the Hatch, Flit and Poetry repos.

Reviewer 2: Ah, interesting. I thought they did.

@jeanas still worth mention the GHA, I think.

Author: Sorry, I forgot about this one, done.

Reviewer 2: Marking this as resolved.

**Never** use ``python setup.py upload`` for this task. In addition to being
:ref:`deprecated <setup-py-deprecated>`, it is insecure.

Integrated workflow tools
=========================

These are tools that combine many features in one command line application, such
as automatically managing virtual environments for a project, building

> distributions, uploading to PyPI, or creating and using lock files.

Reviewer 2: ```suggestion
distributions,
uploading to PyPI, or creating and using tool-specific lock files.
They often call the tools mentioned above under the hood.
```

Reviewer 2: Marking this as resolved.

- :ref:`Hatch`,
- :ref:`Flit`,
- :ref:`PDM`,

> - :ref:`Poetry`.

Reviewer 2: I suppose `Pipenv` can be mentioned in this section too.

:ref:`modernize-setup-py-project` for migration advice).

Lock files
==========

:ref:`pip-tools` and :ref:`Pipenv` are two recognized tools to create lock
files, which contain the exact versions of all packages installed into an
environment, for reproducibility purposes.

Build backends
==============

Popular :term:`build backends <build backend>` for pure-Python packages include,
in alphabetical order:

- Flit-core_ (developed with but separate from :ref:`Flit`). It is meant to be a
  minimal and opinionated build backend. It is not extensible.

- Hatchling_, which is developed along with :ref:`Hatch`, but is separate and can
  be used without Hatch. Hatchling is extensible through a plugin system.

- PDM-backend_ (developed with but separate from :ref:`PDM`). It provides build
  hooks for extensibility.

- Poetry-core_ (developed with but separate from :ref:`Poetry`). It is extensible
  through plugins.

- :ref:`setuptools` (which used to be the only build backend). It can be
  configured using modern standards like the :ref:`[project] table in
  pyproject.toml <writing-pyproject-toml>`, but can also be extended and
  customized via :file:`setup.py`.

>   If you use setuptools, please be aware that it contains many deprecated
>   features which are currently kept for compatibility, but should not be used.

Reviewer 5: ```suggestion
  If you use setuptools, please be aware that some features that predate
  standardisation efforts are now deprecated and only *temporarily kept* for compatibility.
```

in alphabetical order:

- :doc:`Flit-core <flit:pyproject_toml>` -- developed with but separate from :ref:`Flit`.
  A minimal and opinionated build backend. It does not support plugins.

- Hatchling_ -- developed with but separate from :ref:`Hatch`. Supports plugins.

- PDM-backend_ -- developed with but separate from :ref:`PDM`. Supports plugins.

- Poetry-core_ -- developed with but separate from :ref:`Poetry`. Supports
  plugins.

  Unlike other backends on this list, Poetry-core does not support the standard
  :ref:`[project] table <writing-pyproject-toml>` (it uses a different format,
  in the ``[tool.poetry]`` table).

- :ref:`setuptools`, which used to be the only build backend. Supports plugins.

  .. caution::

     If you use setuptools, please be aware that some features that predate
     standardisation efforts are now deprecated and only *temporarily kept*
     for compatibility.

     In particular, do **not** use direct ``python setup.py`` invocations. On the
     other hand, configuring setuptools with a :file:`setup.py` file is still fully
     supported, although it is recommended to use the modern :ref:`[project] table

>      in pyproject.toml <writing-pyproject-toml>` whenever possible and keep

Reviewer 2: I'd like this to avoid ignoring the existence of `setup.cfg`:
```suggestion
     in pyproject.toml <writing-pyproject-toml>` (or :file:`setup.cfg`) whenever possible and keep
```

Tool recommendations
====================

The Python packaging landscape consists of many different tools. For many tasks,
the :term:`Python Packaging Authority <Python Packaging Authority (PyPA)>`
(PyPA, the working group which encompasses many packaging tools and
maintains this guide) purposefully does not make a blanket recommendation; for
example, the reason there are many build backends is that the landscape was
opened up in order to enable the development of new backends serving certain users'
needs better than the previously unique backend, setuptools. This guide does
point to some tools that are widely recognized, and also makes some
recommendations of tools that you should *not* use because they are deprecated
or insecure.

> 

Reviewer 2: Should we mention that some tools focus on many things?
```suggestion

.. tip::

   Note that while most tools are is only mentioned under one
   section in this document, some of them cover several use
   cases.

```

Author: Well, the intent is that all tools presented are single-purpose, except those which are presented under “Integrated workflow tools” (but maybe that title should be more explicit?).

Reviewer 2: Perhaps yes. I also don't fully understand why tox/nox are ripped off of that section. I'd say they belong in it just as well.

Author: I've merged them.

Reviewer 2: Marking as resolved.

Virtual environments
====================

The standard tools to create and use virtual environments manually are
:ref:`virtualenv` (PyPA project) and :doc:`venv <python:library/venv>` (part of
the Python standard library, though missing some features of virtualenv).

Installing packages
===================

:ref:`Pip` is the standard tool to install packages from :term:`PyPI <Python

> Package Index (PyPI)>`. It can install into the global environment or into

Reviewer 2: Also, "virtual" (as a concept/word) is not an opposite of "global". I think this may be confusing to some.

> Package Index (PyPI)>`. It can install into the global environment or into

Reviewer 2: How about "isolated" vs. "shared"?

> Package Index (PyPI)>`. It can install into the global environment or into

Reviewer 2: @jeanas @pfmoore ideas?

> Package Index (PyPI)>`. It can install into the global environment or into

Author: Maybe we should just remove this sentence? Or replace it with “It can install into arbitrary environments.”?

> Package Index (PyPI)>`. It can install into the global environment or into

Reviewer 4: I think remove the sentence. It isn't in the version of the text that's currently live, and I don't think it adds anything.

> Package Index (PyPI)>`. It can install into the global environment or into

Author: Done.

> Package Index (PyPI)>`. It can install into the global environment or into

Reviewer 2: Marking as resolved.

virtual environments. You may want to read pip's recommendations for
:doc:`secure installs <pip:topics/secure-installs>`. Pip is available by default
in most Python installations through the standard library package
:doc:`ensurepip <python:library/ensurepip>`.

Alternatively, consider :ref:`pipx` for the specific use case of installing Python

> command line applications that are distributed through PyPI.

Reviewer 2: I think it's wrong to emphasize CLI apps, while realistically, it is useful in the context of any packages that are considered distributing apps. GUI included, maybe even web(ish) apps with strict pins could be included. So I'd just drop this clarification as not factual.
```suggestion
applications that are distributed through PyPI.
```

Reviewer 2: Alternatively, clarify that there's more to it:
```suggestion
command line (and other types of) applications that are distributed through PyPI.
```

Reviewer 2: I guess that's debatable. But fine. I don't have a better suggestion anyway. Do that.

Reviewer 2: Marking as resolved.

Pipx is a wrapper around pip and venv that installs each
application into a dedicated virtual environment. This avoids conflicts between

> the dependencies of different applications, and also with the system
> (especially on Linux).

Reviewer 2: ```suggestion
the dependencies of different applications, and also with the
ones installed system-wide making use of the same Python
interpreter (especially on GNU/Linux).
```

Reviewer 2: Marking this as resolved.

For scientific software specifically, consider :ref:`Conda` or :ref:`Spack`.

.. todo:: Write a "pip vs. Conda" comparison, here or in a new discussion.

Do **not** use ``easy_install`` (part of :ref:`setuptools`), which is deprecated
in favor of pip (see :ref:`pip vs easy_install` for details). Likewise, do
**not** use ``python setup.py install`` or ``python setup.py develop``, which
are also deprecated (see :ref:`setup-py-deprecated` for background and
:ref:`modernize-setup-py-project` for migration advice).

Lock files
==========

:ref:`pip-tools` and :ref:`Pipenv` are two recognized tools to create lock
files, which contain the exact versions of all packages installed into an
environment, for reproducibility purposes.

Build backends
==============

> 

Reviewer 2: I feel like this might be the place to remind the readers not to treat this list as sorted by popularity.

```suggestion

.. important::

   Please, remember, this document does not seek to steer the reader towards
   a one-stop show solution. It merely enumerates tools observed within the
   packaging community. Different use-cases often need specialized workflows.

```

Author: I'm not sure I understand your meaning of “one-stop shop”. I'm not a native speaker, but I'd read it as “a tool which does everything”, not “a build backend that you should always use”. Also I feel there is already a lot of warning at the beginning of the page.

Reviewer 2: True. Let's wait for more opinions, then. And you're right, I did mean "a tool that does everything". Merely because a lot of people demand one without understanding that it entails. It seems to me that build backends are treated similarly even though, it's a different context. So I kinda wanted to assert the position once again in this section.

Reviewer 2: @pfmoore do you have any thoughts on this?

Reviewer 4: Looks ok to me. I’d say “a particular solution” rather than “a one-stop shop solution”. But otherwise it’s good.

Author: Applied, with a slightly more concise wording.

Popular :term:`build backends <build backend>` for pure-Python packages include,
in alphabetical order:

> - Flit-core_ -- developed with but separate from :ref:`Flit`. A minimal and

Reviewer 2: ```suggestion
- :doc:`Flit-core <flit:pyproject_toml>` -- developed with but separate from :ref:`Flit`. A minimal and
```

Author: Ah, I only see this now. Well, ok, why not.

Reviewer 2: Marking this as resolved.

- Poetry-core_ -- developed with but separate from :ref:`Poetry`. Supports
  plugins.

  Unlike other backends on this list, Poetry-core does not support the standard
  :ref:`[project] table <writing-pyproject-toml>` (it uses a different format,
  in the ``[tool.poetry]`` table).

- :ref:`setuptools`, which used to be the only build backend. Supports plugins.

  If you use setuptools, please be aware that some features that predate
  standardisation efforts are now deprecated and only *temporarily kept*
  for compatibility.

  In particular, do **not** use direct ``python setup.py`` invocations. On the
  other hand, configuring setuptools with a :file:`setup.py` file is still fully
  supported, although it is recommended to use the modern :ref:`[project] table
  in pyproject.toml <writing-pyproject-toml>` whenever possible and keep
  :file:`setup.py` only if programmatic configuration is needed. See
  :ref:`setup-py-deprecated`.

  Other examples of deprecated features you should **not** use include the
  ``setup_requires`` argument to ``setup()`` (use the :ref:`[build-system] table
  <pyproject-guide-build-system-table>` in :file:`pyproject.toml` instead), and

>   the ``easy_install`` command (cf. :ref:`pip vs easy_install`).

Reviewer 2: How about wrapping this with `.. caution::`?

Reviewer 2: Marking this as resolved.

- :ref:`meson-python` -- C, C++, Fortran, Rust, and other languages supported by Meson,
- :ref:`scikit-build-core` -- C, C++, Fortran, and other languages supported by CMake,
- :ref:`maturin` -- Rust, via Cargo.

Building distributions
======================

The standard tool to build :term:`source distributions <source distribution (or
"sdist")>` and :term:`wheels <wheel>` for uploading to PyPI is :ref:`build`.  It
will invoke whichever build backend you :ref:`declared
<pyproject-guide-build-system-table>` in :file:`pyproject.toml`.

Do **not** use ``python setup.py sdist`` and ``python setup.py bdist_wheel`` for
this task. All direct invocations of :file:`setup.py` are :ref:`deprecated
<setup-py-deprecated>`.

If you have :term:`extension modules <extension module>` and want to distribute
wheels for multiple platforms, use :ref:`cibuildwheel` as part of your CI setup
to build distributable wheels.

Uploading to PyPI
=================

The standard tool for this task is :ref:`twine`.

**Never** use ``python setup.py upload`` for this task. In addition to being
:ref:`deprecated <setup-py-deprecated>`, it is insecure.

> Task runners
> ============

Reviewer 2: I think it's more accurate to say
```suggestion
Workflow tools
==============
```

Saying "task runners" reduce them to a GNU/make replacement. But really, there's things like venv management that this skips.

Author: I agree that this could be more precise, but if we make this change then I don't know how to call the section about Flit, Hatch, PDM, Pipenv and Poetry, which is currently called “Integrated workflow tools”. Thoughts?

Reviewer 4: Fundamentally, my thought here is that we're trying too hard to classify the tools available in the ecosystem. We might be better leaving it to the tools themselves to choose how they want to present themselves, and not have any headings at all...

Author: I'm not enthusiastic. The problem that this page solves is “it's too hard to understand how the tools fit together”. If it doesn't classify them, then it doesn't serve its purpose well, IMO.

Reviewer 2: I suppose, I'd be in favor of combining these two sections into one. I feel like there's not enough differences in general to justify outing tox/nox vs. poetry/hatch et al.

Author: It looks like not everyone has the same mental model as me on the nox/tox vs. Flit/Hatch/PDM/Pipenv/Poetry distinction.

For background, the two I've used the most are tox and Hatch, and I've sometimes used nox (in this repo). Tox is an environment manager that lets me define virtual environments and run tasks in them. Meanwhile, Hatch also does environment management and task execution, but also: building, uploading to PyPI, and managing Python interpreters (plus reformatting the code). So I view tox as a single-purpose tool (environment manager) along with other single-purpose tools (build, twine, pyenv) and Hatch as a tool that combines the features of lots of single-purpose tools (build, twine, tox, pyenv).

Merging these two sections is possible but I'm not sure how I'd fill the description.

Making the list task-oriented is also possible, but then Flit, Hatch, Poetry and PDM all are going to get listed in three categories (building, publishing, managing environments, except Flit doesn't manage environments).

This also makes me think that by my categorization, Pipenv should be with tox and nox. (For my defense, I know very little about it except that it sparked controversies years ago.)

Reviewer 2: Technically, tox is used in the same way. People automate calling the underlying tools (like build or twine) and have tox as a workflow tool with a unified interfaces. It doesn't have some things built into the codebase but it also has plugins and stuff. So the use case for it is exactly the same, it just allows more low-level configuration. And it does implement some things that others don't (like parallel execution of several things), but that's just one of the differences. Other tools also differ between each other.

All of these are workflow tools. They just allow different levels of configuration and different have things implemented by default. But from the project contributor (end-user of those tools / the caller), they provide a CLI to call those processes as a part of their workflow.

@gaborbernat correct me if I'm missing something...

Reviewer 6: You're correct.

Author: Ok, I've merged the two sections.

Reviewer 2: Marking as resolved, then.

These tools allow you to define and invoke "tasks", such as running tests,
compiling documentation, regenerating some files, etc. Each task can be executed
in a dedicated virtual environment with the dependencies it requires. The most

> common tools for this are tox_ and nox_.

Reviewer 2: ```suggestion
common tools for this are :doc:`tox <tox:index>` and :doc:`nox <nox:index>`.
```

Reviewer 2: Marking this as resolved.

Integrated workflow tools
=========================

These are tools that combine many features in one command line application, such
as automatically managing virtual environments for a project, building
distributions, uploading to PyPI, or creating and using (tool-specific) lock
files. They often call the tools mentioned above under the hood. In alphabetical
order:

- :ref:`Flit`,
- :ref:`Hatch`,
- :ref:`PDM`,
- :ref:`Pipenv`,
- :ref:`Poetry`.

> .. _flit-core: https://pypi.org/project/flit-core/

Reviewer 2: Flit is wired through intersphinx.

Author: Yes, but I'd like to point to Flit-core here, and preferably not to the main Flit docs, to emphasize the distinction. (The text is `` Flit-core_ -- developed with but separate from :ref:`Flit`. ``.)

Reviewer 2: Yeah, I wasn't completely sure about this one. Though, I see that you still applied the suggested change where I opted to link the doc that talks about `[build-system]` which is closer to the middle ground, it seems.

Marking this as resolved.

.. _hatchling: https://pypi.org/project/hatchling/

> .. _nox: https://nox.thea.codes

Reviewer 2: Nox is wired through intersphinx.

Reviewer 2: Marking this as resolved.

.. _pdm-backend: https://backend.pdm-project.org
.. _poetry-core: https://pypi.org/project/poetry-core/

> .. _tox: https://tox.wiki

Reviewer 2: Tox is wired through intersphinx.

Reviewer 2: Marking this as resolved.

a build system with dedicated support for the language the extension is written in,
for example:

- :ref:`setuptools` -- natively supports C and C++ (with third-party plugins for Go and Rust),
- :ref:`meson-python` -- C, C++, Fortran, Rust, and other languages supported by Meson,
- :ref:`scikit-build-core` -- C, C++, Fortran, and other languages supported by CMake,
- :ref:`maturin` -- Rust, via Cargo.

Building distributions
======================

The standard tool to build :term:`source distributions <source distribution (or
"sdist")>` and :term:`wheels <wheel>` for uploading to PyPI is :ref:`build`.  It
will invoke whichever build backend you :ref:`declared
<pyproject-guide-build-system-table>` in :file:`pyproject.toml`.

Do **not** use ``python setup.py sdist`` and ``python setup.py bdist_wheel`` for
this task. All direct invocations of :file:`setup.py` are :ref:`deprecated
<setup-py-deprecated>`.

If you have :term:`extension modules <extension module>` and want to distribute
wheels for multiple platforms, use :ref:`cibuildwheel` as part of your CI setup
to build distributable wheels.

Uploading to PyPI
=================

> For projects hosted on GitHub, PyPI recommends using :ref:`trusted publishing

Reviewer 2: Maybe, mention it's the PyPA?
```suggestion
For projects hosted on GitHub, the PyPA and the PyPI recommend using :ref:`trusted publishing
```

Reviewer 2: Alternatively,
```suggestion
For projects hosted on GitHub, it is recommended to use the :ref:`trusted publishing
```

Author: I'll go for the second, it's simpler.

Build backends
==============

Popular :term:`build backends <build backend>` for pure-Python packages include,
in alphabetical order:

- Flit-core_ (developed with but separate from :ref:`Flit`). It is meant to be a
  minimal and opinionated build backend. It is not extensible.

- Hatchling_, which is developed along with :ref:`Hatch`, but is separate and can
  be used without Hatch. Hatchling is extensible through a plugin system.

- PDM-backend_ (developed with but separate from :ref:`PDM`). It provides build
  hooks for extensibility.

- Poetry-core_ (developed with but separate from :ref:`Poetry`). It is extensible
  through plugins.

> - :ref:`setuptools` (which used to be the only build backend). It can be configured

Reviewer 5: ```suggestion
- :ref:`setuptools` (which used to be the only build backend). It can be configured
  using modern standards like `pyproject.toml`, but can also be extended
  and supports customisation via `setup.py`.
```

  <pyproject-guide-build-system-table>` of :file:`pyproject.toml` instead), or
  the ``easy_install`` command (cf. :ref:`pip vs easy_install`).

Do **not** use :ref:`distutils`, which is deprecated, and has been removed from
the standard library in Python 3.12, although it still remains available from
setuptools.

For packages with :term:`extension modules <extension module>`, you may use

> setuptools. However, you can also use a build system with dedicated support for
> the language the extension is written in. For example, you could choose Meson or
> CMake for C, C++, Fortran and many other compiled languages; or Cargo for Rust
> specifically. You can then bridge this build system to Python using a dedicated
> build backend:

Reviewer 5: This is probably more direct:

```suggestion
a build system with dedicated support for the language the extension is written in:

- :ref:`setuptools` - natively supports C/C++ (with 3rd-party plugins for Go and Rust);

- :ref:`meson-python` - C, C++, Fortran, and Rust and other languages supported by Meson;

- :ref:`scikit-build-core` - C, C++, Fortran and other languages supported by CMake;

- :ref:`maturin` - Rust, via Cargo.
```
